// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: teacher.proto

package client

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on TeacherFindByUsernameRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *TeacherFindByUsernameRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TeacherFindByUsernameRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// TeacherFindByUsernameRequestMultiError, or nil if none found.
func (m *TeacherFindByUsernameRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *TeacherFindByUsernameRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if l := utf8.RuneCountInString(m.GetTeacherUsername()); l < 4 || l > 30 {
		err := TeacherFindByUsernameRequestValidationError{
			field:  "TeacherUsername",
			reason: "value length must be between 4 and 30 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return TeacherFindByUsernameRequestMultiError(errors)
	}

	return nil
}

// TeacherFindByUsernameRequestMultiError is an error wrapping multiple
// validation errors returned by TeacherFindByUsernameRequest.ValidateAll() if
// the designated constraints aren't met.
type TeacherFindByUsernameRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TeacherFindByUsernameRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TeacherFindByUsernameRequestMultiError) AllErrors() []error { return m }

// TeacherFindByUsernameRequestValidationError is the validation error returned
// by TeacherFindByUsernameRequest.Validate if the designated constraints
// aren't met.
type TeacherFindByUsernameRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TeacherFindByUsernameRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TeacherFindByUsernameRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TeacherFindByUsernameRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TeacherFindByUsernameRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TeacherFindByUsernameRequestValidationError) ErrorName() string {
	return "TeacherFindByUsernameRequestValidationError"
}

// Error satisfies the builtin error interface
func (e TeacherFindByUsernameRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTeacherFindByUsernameRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TeacherFindByUsernameRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TeacherFindByUsernameRequestValidationError{}

// Validate checks the field values on TeacherFindByUsernameResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *TeacherFindByUsernameResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TeacherFindByUsernameResponse with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// TeacherFindByUsernameResponseMultiError, or nil if none found.
func (m *TeacherFindByUsernameResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *TeacherFindByUsernameResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTeacher()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TeacherFindByUsernameResponseValidationError{
					field:  "Teacher",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TeacherFindByUsernameResponseValidationError{
					field:  "Teacher",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTeacher()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TeacherFindByUsernameResponseValidationError{
				field:  "Teacher",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return TeacherFindByUsernameResponseMultiError(errors)
	}

	return nil
}

// TeacherFindByUsernameResponseMultiError is an error wrapping multiple
// validation errors returned by TeacherFindByUsernameResponse.ValidateAll()
// if the designated constraints aren't met.
type TeacherFindByUsernameResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TeacherFindByUsernameResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TeacherFindByUsernameResponseMultiError) AllErrors() []error { return m }

// TeacherFindByUsernameResponseValidationError is the validation error
// returned by TeacherFindByUsernameResponse.Validate if the designated
// constraints aren't met.
type TeacherFindByUsernameResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TeacherFindByUsernameResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TeacherFindByUsernameResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TeacherFindByUsernameResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TeacherFindByUsernameResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TeacherFindByUsernameResponseValidationError) ErrorName() string {
	return "TeacherFindByUsernameResponseValidationError"
}

// Error satisfies the builtin error interface
func (e TeacherFindByUsernameResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTeacherFindByUsernameResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TeacherFindByUsernameResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TeacherFindByUsernameResponseValidationError{}
